File: src/main/java/maestro/process/FollowProcess.java
package maestro.process;

import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import maestro.Maestro;
import maestro.api.pathing.goals.Goal;
import maestro.api.pathing.goals.GoalBlock;
import maestro.api.pathing.goals.GoalComposite;
import maestro.api.pathing.goals.GoalNear;
import maestro.api.pathing.goals.GoalXZ;
import maestro.api.process.IFollowProcess;
import maestro.api.process.PathingCommand;
import maestro.api.process.PathingCommandType;
import maestro.api.utils.BetterBlockPos;
import maestro.utils.MaestroProcessHelper;
import net.minecraft.core.BlockPos;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.item.ItemEntity;
import net.minecraft.world.item.ItemStack;

/**
 * Follow an entity
 *
 * @author leijurv
 */
public final class FollowProcess extends MaestroProcessHelper implements IFollowProcess {

    private Predicate<Entity> filter;
    private List<Entity> cache;
    private boolean into; // walk straight into the target, regardless of settings

    public FollowProcess(Maestro maestro) {
        super(maestro);
    }

    @Override
    public PathingCommand onTick(boolean calcFailed, boolean isSafeToCancel) {
        scanWorld();
        Goal goal = new GoalComposite(cache.stream().map(this::towards).toArray(Goal[]::new));
        return new PathingCommand(goal, PathingCommandType.REVALIDATE_GOAL_AND_PATH);
    }

    private Goal towards(Entity following) {
        BlockPos pos;
        if (Maestro.settings().followOffsetDistance.value == 0 || into) {
            pos = following.blockPosition();
        } else {
            GoalXZ g =
                    GoalXZ.fromDirection(
                            following.position(),
                            Maestro.settings().followOffsetDirection.value,
                            Maestro.settings().followOffsetDistance.value);
            pos = new BetterBlockPos(g.getX(), following.position().y, g.getZ());
        }
        if (into) {
            return new GoalBlock(pos);
        }
        return new GoalNear(pos, Maestro.settings().followRadius.value);
    }

    private boolean followable(Entity entity) {
        if (entity == null) {
            return false;
        }
        if (!entity.isAlive()) {
            return false;
        }
        if (entity.equals(ctx.player())) {
            return false;
        }
        int maxDist = Maestro.settings().followTargetMaxDistance.value;
        if (maxDist != 0 && entity.distanceToSqr(ctx.player()) > maxDist * maxDist) {
            return false;
        }
        return ctx.entitiesStream().anyMatch(entity::equals);
    }

    private void scanWorld() {
        cache =
                ctx.entitiesStream()
                        .filter(this::followable)
                        .filter(this.filter)
                        .distinct()
                        .collect(Collectors.toList());
    }

    @Override
    public boolean isActive() {
        if (filter == null) {
            return false;
        }
        scanWorld();
        return !cache.isEmpty();
    }

    @Override
    public void onLostControl() {
        filter = null;
        cache = null;
    }

    @Override
    public String displayName0() {
        return "Following " + cache;
    }

    @Override
    public void follow(Predicate<Entity> filter) {
        this.filter = filter;
        this.into = false;
    }

    @Override
    public void pickup(Predicate<ItemStack> filter) {
        this.filter = e -> e instanceof ItemEntity && filter.test(((ItemEntity) e).getItem());
        this.into = true;
    }

    @Override
    public List<Entity> following() {
        return cache;
    }

    @Override
    public Predicate<Entity> currentFilter() {
        return filter;
    }
}

File: common/src/main/maestro/process/FollowProcess.kt
package maestro.process

import maestro.Agent
import maestro.api.pathing.goals.Goal
import maestro.api.pathing.goals.GoalBlock
import maestro.api.pathing.goals.GoalComposite
import maestro.api.pathing.goals.GoalNear
import maestro.api.pathing.goals.GoalXZ
import maestro.api.process.IFollowProcess
import maestro.api.process.PathingCommand
import maestro.api.process.PathingCommandType
import maestro.api.utils.PackedBlockPos
import net.minecraft.core.BlockPos
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.item.ItemEntity
import net.minecraft.world.item.ItemStack
import java.util.function.Predicate

/** Follow an entity */
class FollowProcess(
    maestro: Agent,
) : MaestroProcessHelper(maestro),
    IFollowProcess {
    private var filter: Predicate<Entity>? = null
    private var cache: List<Entity>? = null
    private var into: Boolean = false // Walk straight into the target, regardless of settings

    override fun onTick(
        calcFailed: Boolean,
        isSafeToCancel: Boolean,
    ): PathingCommand {
        scanWorld()
        val currentCache = cache ?: emptyList()
        val goal = GoalComposite(*currentCache.map { towards(it) }.toTypedArray())
        return PathingCommand(goal, PathingCommandType.REVALIDATE_GOAL_AND_PATH)
    }

    private fun towards(following: Entity): Goal {
        val pos: BlockPos =
            if (Agent.settings().followOffsetDistance.value == 0.0 || into) {
                following.blockPosition()
            } else {
                val g =
                    GoalXZ.fromDirection(
                        following.position(),
                        Agent.settings().followOffsetDirection.value,
                        Agent.settings().followOffsetDistance.value,
                    )
                PackedBlockPos(g.x, following.position().y.toInt(), g.z).toBlockPos()
            }

        return if (into) {
            GoalBlock(pos)
        } else {
            GoalNear(pos, Agent.settings().followRadius.value)
        }
    }

    private fun followable(entity: Entity?): Boolean {
        if (entity == null) {
            return false
        }

        if (!entity.isAlive) {
            return false
        }

        if (entity == ctx.player()) {
            return false
        }

        val maxDist = Agent.settings().followTargetMaxDistance.value
        val player = ctx.player() ?: return false

        if (maxDist != 0 && entity.distanceToSqr(player) > maxDist * maxDist) {
            return false
        }

        return ctx.entitiesStream().anyMatch { it == entity }
    }

    private fun scanWorld() {
        val currentFilter = filter
        cache =
            if (currentFilter != null) {
                ctx
                    .entitiesStream()
                    .filter { followable(it) }
                    .filter { currentFilter.test(it) }
                    .distinct()
                    .toList()
            } else {
                emptyList()
            }
    }

    override fun isActive(): Boolean {
        if (filter == null) {
            return false
        }

        scanWorld()
        return !cache.isNullOrEmpty()
    }

    override fun onLostControl() {
        filter = null
        cache = null
    }

    override fun displayName0(): String = "Following $cache"

    override fun follow(filter: Predicate<Entity>) {
        this.filter = filter
        this.into = false
    }

    override fun pickup(filter: Predicate<ItemStack>) {
        this.filter =
            Predicate { e ->
                e is ItemEntity && filter.test(e.item)
            }
        this.into = true
    }

    override fun following(): List<Entity>? = cache

    override fun currentFilter(): Predicate<Entity>? = filter
}
